airflow/providers/qdrant/__init__.py

```diff
@@ -14,3 +14,29 @@
 # KIND, either express or implied.  See the License for the
 # specific language governing permissions and limitations
 # under the License.
+#
+# NOTE! THIS FILE IS AUTOMATICALLY GENERATED AND WILL BE
+# OVERWRITTEN WHEN PREPARING DOCUMENTATION FOR THE PACKAGES.
+#
+# IF YOU WANT TO MODIFY THIS FILE, YOU SHOULD MODIFY THE TEMPLATE
+# `PROVIDER__INIT__PY_TEMPLATE.py.jinja2` IN the `dev/breeze/src/airflow_breeze/templates` DIRECTORY
+#
+from __future__ import annotations
+
+import packaging.version
+
+__all__ = ["__version__"]
+
+__version__ = "1.0.0"
+
+try:
+    from airflow import __version__ as airflow_version
+except ImportError:
+    from airflow.version import version as airflow_version
+
+if packaging.version.parse(packaging.version.parse(airflow_version).base_version) < packaging.version.parse(
```

Comment:
I always confused by this long construction, I guess it comes from the fact that we have to lax requirement to `packaging`

https://github.com/apache/airflow/blob/9a10085bb142df8a7699cd66908f1896ae10b5bb/pyproject.toml#L120-L120

`packaging` 14 was released in 2018: https://pypi.org/project/packaging/#history
So I guess we could bump min version to something new, e.g. >= 22 or eve 23, and one day replace this to 

```python
from packaging.version import Version

...

if Version(Version(airflow_version).base_version) < Version("2.6.0"):
   raise ...
```

Reply:
Good idea. the >=22 sounds good. The (small) problem is that we will have to add packaging >=22 to the providers automatically when we release it with this change, but that should be no problem 

It's a matter of changing https://github.com/apache/airflow/blob/main/dev/breeze/src/airflow_breeze/templates/PROVIDER__INIT__PY_TEMPLATE.py.jinja2 (for `__init__.py`) and https://github.com/apache/airflow/blob/main/dev/breeze/src/airflow_breeze/templates/pyproject_TEMPLATE.toml.jinja2 (for package dependencies). I can add shortly

Reply:
I mean we could switch to new format when all providers have min Airflow 2.9.
And until then use current template

Reply:
I think we could also add packaging directly as dependency to all providers. In fact they are already depending on it implicitly (through aiflow) so having >=22 in all of them will only make sure that newer version is installed.

Reply:
I'm also not sure about single version of the truth about where we should store Airflow Version.
I'm have personal filling that we do not need to store it into the `airflow.__version__`,

For internal usage we could import it elsewhere, e.g. airflow.airflow_version
For build, it doesn't matter from which module we would import it.
Third Parties should use [`importlib.metadata.version`](https://docs.python.org/3/library/importlib.metadata.html#distribution-versions) or backports rather.

But this discussion for more wide audience, I will raise it into dev list.

Reply:
`airflow.__version__` is the de-facto standard - many packages are using it. And even our hatch build relies on it because version in pyproject.toml is dynamic and hatchling by default reads it from there.

Reply:
It's simply way, way faster check than using importlib.

Reply:
> airflow.__version__ is the de-facto standard - many packages are using it.

I do not suggest to entirely remove it, at least until Airflow 3, rather than do not use as main way.

Good example, hatch do not define in their codebase __version__ in package, some other also force to use importlib instead, because it is not mandatory by any PEPs.

> It's simply way, way faster check than using importlib.

Internally it doesn't matter where we store/get version until it backward compatible. For third parties importlib potentially should be faster if compare to initialize Airflow settings.

Reply:
We use it with providers already :) - it's not only "internally" - the airflow.__version__ check by providers is really fastest way - and providers are imported after airflow is already initialized, so there is no penalty. If only as interface for providers, it's rather handy

Reply:
I guess we talk about different things. 

My point here that we should change source of the truth for the version to the `airflow.version` module, and use it from there in providers, hatch builds and elsewhere in core.

Keep it for backward compatibility in `airflow.__init__`, e.g. lazy loaded from `airflow.version.`.

This might open opportunity to add addition helpers for prevent common constructions like

```python
packaging.version.parse(packaging.version.parse(airflow_version).base_version
```

or

```python
Version(Version(airflow_version).base_version)
```

And turn it into the 

```python
from airflow.version import base_version

...

if base_version < Version("2.7.0"):
    raise ...
```


The straightforward implementation would be

```diff
diff --git a/airflow/__init__.py b/airflow/__init__.py
index 4af74a44d2..c2b4c84702 100644
--- a/airflow/__init__.py
+++ b/airflow/__init__.py
@@ -26,10 +26,6 @@ isort:skip_file
 """
 from __future__ import annotations
 
-__version__ = "2.9.0.dev0"
-
-# flake8: noqa: F401
-
 import os
 import sys
 from typing import Callable
@@ -82,6 +78,7 @@ __lazy_imports: dict[str, tuple[str, str, bool]] = {
     "Dataset": (".datasets", "Dataset", False),
     "XComArg": (".models.xcom_arg", "XComArg", False),
     "version": (".version", "", False),
+    "__version__": (".version", "__version__", False),
     # Deprecated lazy imports
     "AirflowException": (".exceptions", "AirflowException", True),
 }
@@ -138,3 +135,4 @@ if STATICA_HACK:  # pragma: no cover
     from airflow.models.dag import DAG
     from airflow.models.dataset import Dataset
     from airflow.models.xcom_arg import XComArg
+    from airflow.version import __version__
diff --git a/airflow/version.py b/airflow/version.py
index 38fb013244..c3a1064297 100644
--- a/airflow/version.py
+++ b/airflow/version.py
@@ -17,7 +17,20 @@
 # under the License.
 from __future__ import annotations
 
-# Compat -- somethings access `airflow.version.version` directly
-from airflow import __version__ as version
+__all__ = ["version", "__version__"]
 
-__all__ = ["version"]
+__version__ = "2.9.0.dev0"
+version = __version__
+
+
+def __getattr__(name: str):
+    # PEP-562
+    if name != "base_version":
+        raise AttributeError(f"module {__name__!r} has no attribute {name!r}")
+
+    from packaging.version import Version
+
+    val = Version(Version(__version__).base_version)
+    # Store for next time
+    globals()[name] = val
+    return val
diff --git a/pyproject.toml b/pyproject.toml
index d45c31e03e..29df73e952 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -1221,7 +1221,7 @@ description = "Environment with Python 3.11. No devel installed"
 features = []
 
 [tool.hatch.version]
-path = "airflow/__init__.py"
+path = "airflow/version.py"
 
 [tool.hatch.build.targets.wheel.hooks.custom]
 path = "./hatch_build.py"
```

---

For third parties more convenient and safe way would be importlib.
There is not something new (in general, not about airflow) when a package distribution has a different version rather than defined in `__version__`

In case of Airflow it would be even faster

```console
❯ time python -c "from importlib import metadata; version = metadata.version('apache-airflow'); print(version)"
2.9.0.dev0
python -c   0.06s user 0.03s system 50% cpu 0.175 total

❯ time python -c "from airflow import __version__ as version; print(version)"
2.9.0.dev0
python -c "from airflow import __version__ as version; print(version)"  0.43s user 0.14s system 63% cpu 0.893 total

❯ export _AIRFLOW__AS_LIBRARY=1
❯ time python -c "from airflow import __version__ as version; print(version)"
2.9.0.dev0
python -c "from airflow import __version__ as version; print(version)"  0.37s user 0.11s system 68% cpu 0.706 total
```

Reply:
Yes. We could do that and do it even now, but it will be usable in providers where min_airflow_version is 2.9

Reply:
BTW. I think it's the 3rd or forth iteration on how we store Airflow version.

Reply:
> BTW. I think it's the 3rd or forth iteration on how we store Airflow version.

That is why I would like to open discussion, because someone might have filling: "Lets stop to do this kind of stuff". I've planed initially that I've did it today, but a bit hectic day, so I hope I would have a time and prepare some options and pros and cons.

> Yes. We could do that and do it even now, but it will be usable in providers where min_airflow_version is 2.9

Of we could even use conditions in jinja templates based on the version, configure it one time and when the time is coming for min 2.9 it would replaced automatically.

Reply:
Not really, I works on module level, e.g. you could add version for each separate module which use in this marker, I'm not sure that someone really crazy to do that, but PEP8 allow that. 

But even there check cases when version not defined.

```python
    verattr = getattr(mod, "__version__", None)
    if minversion is not None:
        # Imported lazily to improve start-up time.
        from packaging.version import Version

        if verattr is None or Version(verattr) < Version(minversion):
            raise Skipped(
                "module %r has __version__ %r, required is: %r"
                % (modname, verattr, minversion),
                allow_module_level=True,
            )
```

Reply:
[PEP 396 – Module Version Numbers](https://peps.python.org/pep-0396/) - Status Rejected

> This PEP was formally rejected on 2021-04-14. The packaging ecosystem has changed significantly in the intervening years since this PEP was first written, and APIs such as importlib.metadata.version() [[11]](https://peps.python.org/pep-0396/#id14) provide for a much better experience.

Reply:
Something new for me that it even discussed to make it standard.